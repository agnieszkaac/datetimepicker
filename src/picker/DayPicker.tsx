import React from "react";
import cx from "classnames";
import moment from "moment";

import { DayPickerProps, DayProps } from "../types";

const Day: React.FunctionComponent<DayProps> = ({
  value,
  selected,
  isToday,
  onPick,
}) => (
  <button
    className={cx("day", { selected: selected, "is-today": isToday })}
    onClick={onPick}
    value={value}
  >
    {value}
  </button>
);

export const DayPicker: React.FunctionComponent<DayPickerProps> = ({
  value,
  monthLength,
  onPick,
}) => (
  <div className="day-picker">
    {Array.from({ length: monthLength }, (e, i) => (
      <Day
        key={i + 1}
        value={(i + 1).toString()}
        selected={value === (i + 1).toString()}
        isToday={moment().date().toString() === value}
        onPick={onPick}
      />
    ))}
  </div>
);
