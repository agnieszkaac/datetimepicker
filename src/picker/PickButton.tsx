import React from "react";
import { Moment } from "moment";

export type PickButtonProps = {
  date: Moment;
  format: string;
  className?: string;
  onClick: (e: React.MouseEvent<HTMLButtonElement>) => void;
};

export const PickButton: React.FC<PickButtonProps> = (props) => {
  const { date, format, className, onClick } = props;
  return (
    <button
      type="button"
      value={date.toString()}
      className={className}
      onClick={onClick}
    >
      {date.format(format)}
    </button>
  );
};
