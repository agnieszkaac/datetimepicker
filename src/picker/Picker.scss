.picker {
  display: flex;
  flex-flow: column;
  justify-content: center;
  align-items: center;
  min-width: 200px;
  padding: 4px;
  margin-top: 4px;
  background-color: white;
  border: 1px solid #e2e2e2;
  border-radius: 2px;
}
