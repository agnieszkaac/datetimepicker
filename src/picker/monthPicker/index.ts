export * from "./Month";
export * from "./MonthPicker";