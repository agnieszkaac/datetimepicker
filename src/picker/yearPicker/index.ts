import { Year, YearProps } from "./Year";
import { YearPicker, YearPickerProps } from "./YearPicker";

export { Year, YearPicker, YearProps, YearPickerProps };
