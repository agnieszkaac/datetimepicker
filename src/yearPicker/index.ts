import { Year } from "./Year";
import { YearPicker } from "./YearPicker";
import { YearProps, YearPickerProps } from "./types";

export { Year, YearPicker, YearProps, YearPickerProps };
