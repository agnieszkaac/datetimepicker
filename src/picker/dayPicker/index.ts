import { Day } from "./Day";
import DayPicker from "./DayPicker";
import { DayProps, DayPickerProps } from "./types";

export { Day, DayPicker, DayProps, DayPickerProps };
