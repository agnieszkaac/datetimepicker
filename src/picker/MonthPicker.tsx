import React from "react";
import moment from "moment";
import cx from "classnames";

import { PickerComponentProps } from "./types";
import { defaultGridSize as length } from "./utils";
import { DateContext, ViewContext } from "../state";

import { PickButton } from "./PickButton";

import "./MonthPicker.scss";

export const MonthPicker: React.FC<PickerComponentProps> = ({ onPick }) => {
  const { viewDate } = React.useContext(ViewContext);
  const { date } = React.useContext(DateContext);

  return (
    <div className="month-picker">
      {Array.from({ length }, (_e, i) => {
        const month = moment(viewDate).month(i);
        return (
          <PickButton
            key={i}
            type="month"
            date={month.startOf("month")}
            format="MMM"
            className={cx({
              selected: date?.month() === i && date?.isSame(viewDate, "y"),
              "is-today": month.isSame(moment(), "M"),
            })}
            onClick={onPick}
          />
        );
      })}
    </div>
  );
};
