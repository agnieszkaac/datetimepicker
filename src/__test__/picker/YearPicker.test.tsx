import React from "react";
import MockDate from "mockdate";
import moment, { Moment } from "moment";
import { render } from "@testing-library/react";

import { YearPicker, YearPickerProps } from "../../picker/yearPicker";

describe("<YearPicker /> should", () => {
  let date: Moment | undefined;
  let viewDate: Moment;
  const onPick = jest.fn((_e) => {});
  let props: YearPickerProps;

  beforeEach(() => {
    MockDate.set("2020-01-01");

    date = undefined;
    viewDate = moment();
    props = {
      date,
      viewDate,
      onPick,
    };
  });

  afterEach(() => {
    MockDate.reset();
  });

  it("match snapshot", () => {
    const { asFragment } = render(<YearPicker {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("render 12 years in a grid", () => {
    const { getAllByTestId } = render(<YearPicker {...props} />);
    const years = getAllByTestId("year");

    expect(years.length).toEqual(12);
  });

  it("render selected year when date prop is passed and within viewDate", () => {
    const { getByText, rerender } = render(<YearPicker {...props} />);
    const year = getByText("2022");

    expect(year).not.toHaveClass("selected");

    props.date = moment("2022-01-01");
    rerender(<YearPicker {...props} />);

    expect(year).toHaveClass("selected");
  });

  it("not render selected year when date is not within viewDate", () => {
    props.date = moment();
    props.viewDate = moment("2050-01-01");

    const { queryByText } = render(<YearPicker {...props} />);
    const year = queryByText("2020");

    expect(year).toBeNull();
  });

  it("render current year when within viewDate", () => {
    const { getByText } = render(<YearPicker {...props} />);
    const year = getByText("2020");

    expect(year).toHaveClass("is-today");
  });

  it("not render current year when not within viewDate", () => {
    props.viewDate = moment("2050-01-01");

    const { queryByText } = render(<YearPicker {...props} />);
    const year = queryByText("2020");

    expect(year).toBeNull();
  })
});
