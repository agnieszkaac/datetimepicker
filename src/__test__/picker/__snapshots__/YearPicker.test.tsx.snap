// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<YearPicker /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="year-picker"
    data-testid="year-picker"
  >
    <button
      class="year is-today"
      data-testid="year"
      type="button"
      value="2020-01-01T00:00:00.000Z"
    >
      2020
    </button>
    <button
      class="year"
      data-testid="year"
      type="button"
      value="2021-01-01T00:00:00.000Z"
    >
      2021
    </button>
    <button
      class="year"
      data-testid="year"
      type="button"
      value="2022-01-01T00:00:00.000Z"
    >
      2022
    </button>
    <button
      class="year"
      data-testid="year"
      type="button"
      value="2023-01-01T00:00:00.000Z"
    >
      2023
    </button>
    <button
      class="year"
      data-testid="year"
      type="button"
      value="2024-01-01T00:00:00.000Z"
    >
      2024
    </button>
    <button
      class="year"
      data-testid="year"
      type="button"
      value="2025-01-01T00:00:00.000Z"
    >
      2025
    </button>
    <button
      class="year"
      data-testid="year"
      type="button"
      value="2026-01-01T00:00:00.000Z"
    >
      2026
    </button>
    <button
      class="year"
      data-testid="year"
      type="button"
      value="2027-01-01T00:00:00.000Z"
    >
      2027
    </button>
    <button
      class="year"
      data-testid="year"
      type="button"
      value="2028-01-01T00:00:00.000Z"
    >
      2028
    </button>
    <button
      class="year"
      data-testid="year"
      type="button"
      value="2029-01-01T00:00:00.000Z"
    >
      2029
    </button>
    <button
      class="year"
      data-testid="year"
      type="button"
      value="2030-01-01T00:00:00.000Z"
    >
      2030
    </button>
    <button
      class="year"
      data-testid="year"
      type="button"
      value="2031-01-01T00:00:00.000Z"
    >
      2031
    </button>
  </div>
</DocumentFragment>
`;
